### ISSUE TITLE
404 at https://www.google.com:443//recaptcha/api/verify

### ISSUE THREAD

**Issue by r-martins** [ProxiBlue/reCaptcha#46]
When coming back and editing a previous section using Firecheckout, the module returns an alert of Invalid Captcha.

Debuging the module, I found that it gets a 404 when doing the following request to https://www.google.com:443//recaptcha/api/verify.

```php
    private function _sendRequest($path, $params)
    {
        $httpRequest = new Zend_Http_Client(
            ProxiBlue_ReCaptcha_Helper_Data::RECAPTCHA_API_SERVER
            . '/'
            . ProxiBlue_ReCaptcha_Helper_Data::RECAPTCHA_API_PATH
            . '/'
            . $path
        );
        $httpRequest->setParameterPost(array_merge(array('remoteip' => $_SERVER['REMOTE_ADDR']), $params));
        $response = $httpRequest->request('POST');
        if ($response->getStatus() != 200) {
            $this->_debug('Bad response from captcha gateway. we got ' . $response->getStatus());
            $this->_debug('$httpRequest => ' . print_r($httpRequest,true));
            Mage::throwException('Bad response from captcha gateway. we got ' . $response->getStatus());
        }

        return $response->getBody();

    }
```

I'm still debugging, but just wondering if that is really the correct url.

**Comment by InsomniumBR** [comment]
(Fala Ricardo! ;) nos encontrando por aí - @WagnerBertolini)

@r-martins I was getting the same response code because I had not configured the api keys from google.

Other things I have done to make it work properly on my 1.9.3.4 store:

- copying the captcha.xml even for rwd (instructions on readme)
- removing ProxiBlue folder and reinstalling the old version 2.0.1 of this plugin.

Now it works. =]

**Comment by ProxiBlue** [comment]
@InsomniumBR 

To confirm, you are also on Firecheckout?

regading:

> removing ProxiBlue folder and reinstalling the old version 2.0.1 of this plugin.

Can you confirm that you are then running m1< 1.9.4 without security patch SUPEE 10975?

Any version past 2.0.1 will not work in 1.9.4 or prior with said patch.

https://github.com/ProxiBlue/reCaptcha#versions

**Comment by ProxiBlue** [comment]
@r-martins - sorry for teh late response.

> I'm still debugging, but just wondering if that is really the correct url.

No, it does not look correct.

**Comment by InsomniumBR** [comment]
@ProxiBlue 

I am running 1.9.3.4 and I think I did not apply the security patch SUPEE 10975 yet.

And I am sorry! Not using Firecheckout.

**Comment by ProxiBlue** [comment]
> I'm still debugging, but just wondering if that is really the correct url.

Nope, it is correct:

as per their docs:

https://developers.google.com/recaptcha/docs/verify

API Request
URL: https://www.google.com/recaptcha/api/siteverify METHOD: POST

I suspect you will get a 404 if any of the required params are not populated:

**Comment by ProxiBlue** [comment]
And again, ahhh - no, I I call:

https://www.google.com/recaptcha/api/siteverify

direct, it works with a json result.

The url generated by the module is not working.

The double slash maybe // ?

**Comment by ProxiBlue** [comment]
```
siteverify
```

not verify

so yes, I stand corrected, the generated URL is incorrect.

I am sure I fixed this, as it sound familiar...

**Comment by ProxiBlue** [comment]
Ok, the issue is here: 

https://github.com/ProxiBlue/reCaptcha/blob/master/app/code/community/ProxiBlue/ReCaptcha/Model/Recaptcha.php#L133

That is a fallback routine, which tries to rebuilt the request response is not in params.
If this happens, should really try determine why.

In most cases it will allow the verification to continue, or just pass (an improvement on v3 one day, maybe)

The call there is using the old 2.0.1 code constant, and should be using the new one

**Comment by ProxiBlue** [comment]
After that change, RECAPTCHA_VERIFY_PATH is no longer in use, so the constnat is now dropped as well.

**Comment by ProxiBlue** [comment]
@r-martins 

Likely resolved in 2.3.2 release. Not tested, as I don;t have time right now to try and reproduce that fallback scenario form happening.

**Comment by r-martins** [comment]
@ProxiBlue Thanks a lot for the reply.
In fact the siteverify was the problem regarding the 404 error.
However, when we come back to a previous step in Firecheckout, I guess we need to add something else to trigger the reValidation, becaus now the siteverify returns `{
  "success": false,
  "error-codes": [
    "timeout-or-duplicate"
  ]
}` (incorrect captcha).

I will try the new version now.

**Comment by r-martins** [comment]
Just tested the 2.3.2 version.
The above problem (timeout or duplicate) persists, and the 404 too (as the `RECAPTCHA_VERIFY_PATH` constant is still there).
I did a quick video where:
* the fist request works, and the captcha is successful
* the second fail due to duplicate
* retrying returns 404

[![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/4FI2MD6X89o/0.jpg)](https://www.youtube.com/watch?v=4FI2MD6X89o)

@ProxiBlue is there any Js method I could trigger to re-verify the captcha with new credentials?

**Comment by ProxiBlue** [comment]
@r-martins - Yes, you are correct. something went wrong making that release. I been doing to much at once, lost track of a step.

New release: 2.3.3 which the change now.

https://github.com/ProxiBlue/reCaptcha/releases/tag/2.3.3

**Comment by ProxiBlue** [comment]
> is there any Js method I could trigger to re-verify the captcha with new credentials

Try add this to start of each buttons actions.

https://github.com/ProxiBlue/reCaptcha/blob/master/app/design/frontend/base/default/template/captcha/recaptcha.phtml#L127

**Comment by r-martins** [comment]
Thanks for the update @ProxiBlue .
Didn't work for me.
When adding a manual breakpoint and trigger ir manually, it processes something but the id with the captcha is not updated.
When adding it automatically (on the trigger), I get a timeout error..
![Finalize_sua_inscrição_e_comece_a_assistir_agora](https://user-images.githubusercontent.com/191149/85975326-18bacc80-ba1b-11ea-98d2-d188fdad68ab.jpg)

**Comment by r-martins** [comment]
My mistake. `gerecaptcha.execute()` was being called before the reset and the loop.
It works!!
Thanks a lot one more time, @ProxiBlue.
Let me buy you a coffee when you come to nsw. ;)

**Comment by r-martins** [comment]
@ProxiBlue fixed the 404 on 2.3.3
